spyderlib/otherplugins.py

```diff
@@ -8,52 +8,140 @@
+
+    if module_name in modnames:
+        return
+    try:
+        # First add a mock module with the LOCALEPATH attribute so that the
+        # helper method can fin the locale on import
+        mock = _ModuleMock()
+        mock.LOCALEPATH = Path(namespace_path) / name / 'locale'
+        sys.modules[module_name] = mock
+
+        if PY33:
+            loader = importlib.machinery.PathFinder.find_module(
+                name, [namespace_path])
+            module = loader.load_module(name)
+        elif PY3:
+            spec = importlib.machinery.PathFinder.find_spec(name,
+                                                            [namespace_path])
+            module = spec.loader.load_module(name)
+        else:
+            info = imp.find_module(name, [namespace_path])
+            module = imp.load_module(module_name, *info)
+
+        # Then restore the actual loaded module instead of the mock
+        sys.modules[module_name] = module
+
+        modlist.append(module)
+        modnames.append(module_name)
+    except Exception:
+        sys.stderr.write("ERROR: 3rd party plugin import failed for "
+                         "`{0}`\n".format(module_name))
+        traceback.print_exc(file=sys.stderr)
+
+
+def create_userplugins_files(path):
+    """
+    """
+    if not path.isdir():
+        path.makedirs_p()
+
+    init_file = "__init__.py"
+    init_file_content = """# -*- coding: utf-8 -*-
+'''
+'spyplugins' makes uses of namespace packages to keep different plugins
+organized in the sitepackages directory and in the user directory.
+
+Spyder plugins can be of 'io' type or 'ui' type. Each type also makes use
+of namespace packages.
+
+For more information on namespace packages visit:
+- https://www.python.org/dev/peps/pep-0382/
+- https://www.python.org/dev/peps/pep-0420/
+'''
+
+# Declare as a namespace package
+__import__('pkg_resources').declare_namespace(__name__)
+"""
+    new_path = path / init_file
+    if not (new_path.isfile() and new_path.text == init_file_content):
+        new_path.write_text(init_file_content)
+
+
+def get_spyderplugins_mods(io=False):
+    """Import modules from plugins package and return the list"""
+    base_namespace = "spyplugins"
+
+    if io:
+        plugins_namespace = "io"
+    else:
+        plugins_namespace = "ui"
+
+    namespace = '.'.join([base_namespace, plugins_namespace])
+
+    # Import parent module
+    importlib.import_module(namespace)
+
+    # Create user directory
+    user_conf_path = Path(get_conf_path())
+    user_plugin_basepath = user_conf_path / base_namespace
+    user_plugin_path = user_conf_path / base_namespace / plugins_namespace
+
+    create_userplugins_files(user_plugin_basepath)
+    create_userplugins_files(user_plugin_path)
+
+    modlist, modnames = [], []
+
+    # The user plugins directory is given the priority when looking for modules
+    for plugin_path in [user_conf_path] + sys.path:

```

Comment:
They could be installed anywhere on python's path theoretically. For example by using the `--user` option from pip it would be in `~/.local`